CI: Rebase NumPy compiled extension test modules on Cygwin

DWesl opened the pull request:
This seems to be causing fork failures on Cygwin.  Let's see if this reduces the number of failures.

Fixes #23070?
<!--         ----------------------------------------------------------------
                MAKE SURE YOUR PR GETS THE ATTENTION IT DESERVES!
                ----------------------------------------------------------------

*  FORMAT IT RIGHT:
      https://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message

*  IF IT'S A NEW FEATURE OR API CHANGE, TEST THE WATERS:
      https://www.numpy.org/devdocs/dev/development_workflow.html#get-the-mailing-list-s-opinion

*  HIT ALL THE GUIDELINES:
      https://numpy.org/devdocs/dev/index.html#guidelines

*  WHAT TO DO IF WE HAVEN'T GOTTEN BACK TO YOU:
      https://www.numpy.org/devdocs/dev/development_workflow.html#getting-your-pr-reviewed
-->

DWesl committed:
CI: Split up NumPy compiled extension test modules

This seems to be causing fork failures on Cygwin.  Let's see if this reduces the number of failures.

DWesl commented:
Four hours to see if all extension modules can find their dependencies.  That step really shouldn't take longer than the test that loads all the modules, and definitely not more than twice as long.

DWesl committed:
CI: Use -k to split test cases instead of --ignore

--ignore led to more failures, which is not ideal.

DWesl committed:
CI: Split tests by submodule to see if that works.

DWesl commented:
Six hundred errors on the last two attempts, which is less than ideal given that the original problem was eight.

DWesl committed:
FIX: Make sure CI fails if tests fail.

DWesl committed:
CI: Print more debug information while rebasing.

Let's see if this shows why everything's got a fork() failure now.

DWesl committed:
CI: Put timeouts on Cygwin dependency checks.

It should be plenty of time; each of these commands should complete in maybe five seconds per file on a slow day.

DWesl committed:
TST: Rebase F2Py test modules on Cygwin.

Let's see if this fixes the 8-50 fork failures.

seberg commented:
On the cygwin help it still mentions to check for bad "bloda" software running (mostly anti-virus stuff), no idea if that can help...

But besides that, I am wondering if we have to opt for some hot-fix to get CI green again soon.

DWesl commented:
> On the cygwin help it still mentions to check for bad "bloda" software running (mostly anti-virus stuff), no idea if that can help...
> 

It can, but I don't see anything obvious:
https://github.com/actions/runner-images/blob/main/images/win/Windows2022-Readme.md

> But besides that, I am wondering if we have to opt for some hot-fix to get CI green again soon.

```python
pytestmark = pytest.mark.xfail(sys.platform == "cygwin", message="Random fork() failures", error=BlockingIOError)
```
at module level in a couple of places *should* work to get CI green, but I haven't tested that.

DWesl committed:
FIX: Add glob import for test module rebase.

Forgot to check this earlier.

DWesl committed:
CI: Run each F2Py test in a separate process

Hopefully this will keep the main memory space clear enough to allow all tests to succeed.

DWesl committed:
CI: Split F2Py and non-F2Py tests again

This should help debugging a bit.

DWesl commented:
The equivalent of `-j 4` may separate the extension modules enough to work without the overhead of `--forked`

DWesl committed:
CI: Run F2Py tests in parallel on Cygwin

Hopefully this helps, since --forked takes almost five hours.

DWesl committed:
Revert "TST: Rebase F2Py test modules on Cygwin."

This reverts commit 608864613b801b9c85573186a9d07eeac5e7e465.

DWesl committed:
Revert "FIX: Add glob import for test module rebase."

This reverts commit 33709afdbbc47b7adb7dd06a730246d8c02f724f.

DWesl committed:
CI: Increase number of processes for F2Py tests

Let's see if this eliminates the fork failures.

It's back to around where it was before I started tinkering, so this approach may not work.

DWesl committed:
CI: Revert increase in parallel test processes.

Tests hang, which is less than ideal.
Hopefully this one works well.

DWesl commented:
The tests for every submodule but `f2py` are working, it's just that last submodule causing problems.  We could drop those tests from CI runs, or I could port over the F2Py xfail list from #23114 and see if the combination allows CI to pass.

DWesl commented:
> > On the cygwin help it still mentions to check for bad "bloda" software running (mostly anti-virus stuff), no idea if that can help...
> 
> It can, but I don't see anything obvious: https://github.com/actions/runner-images/blob/main/images/win/Windows2022-Readme.md

I found the image description:
https://github.com/actions/runner-images/blob/win22/20230123.1/images/win/Windows2022-Readme.md
[The changes on Jan 24](https://github.com/actions/runner-images/blob/win22/20230123.1/images/win/Windows2022-Readme.md) (vcpkg, Kubectl, and Pulumi are the ones that catch my attention, the others don't look relevant) don't have obvious antivirus or DLL-handling changes.
[The changes on Jan 19](https://github.com/actions/runner-images/commit/23a732a107efcd2a92097c929faa797515dd60ed) (Vcpkg and Pester are the ones I don't know) also doesn't show anything obvious.
[The Jan 12 changes](https://github.com/actions/runner-images/commit/a22d921c177b4ff25548bd51282a72a0ffce9bed) are probably too far back to be relevant to this, since [it got noticed on Jan 22](https://github.com/numpy/numpy/pull/23066#issuecomment-1399618055)

> 
> > But besides that, I am wondering if we have to opt for some hot-fix to get CI green again soon.
> 
> ```python
> pytestmark = pytest.mark.xfail(sys.platform == "cygwin", message="Random fork() failures", error=BlockingIOError)
> ```
> 
> at module level in a couple of places _should_ work to get CI green, but I haven't tested that.

#23114 now up: it takes a few more places than I thought it would if running all the tests together.

DWesl committed:
TST: Rebase F2Py-built extension modules.

Also adjust CI so they don't immediately collide with NumPy.
I forgot to do that last time, which caused problems.

DWesl commented:
Or write the locations of the NumPy extension modules to the rebase database so the F2Py tests don't put their modules right on top of those.

mattip commented on .github/workflows/cygwin.yml:
```diff
@@ -65,9 +65,21 @@ jobs:
+        # Not sure if that will run the top-level tests.
+        shell: "C:\\tools\\cygwin\\bin\\bash.exe -o igncr -eo pipefail {0}"
+        run: |
+          for submodule in array_api compat core distutils fft lib linalg ma matrixlib polynomial random tests typing;
```
I am not sure what is the best path forward here:
- a list of directories to use, which will break the next time we add/remove a top-level directory
- some ugly path parsing like `find numpy -maxdepth 1 -type d | cut -f2 -d/ | grep -v f2py | grep -v numpy | grep -v "^_"` to get all the directories except `f2py`
- add a command line switch to `runtests.py` to allow skipping a subdirectory
- move the build to meson, and add a command line option to dev.py to skip a subdirectory

What you have is the simplest, but will be fragile. On the other hand, how often do we change the top-level directories?

Maybe is there a way to verify that your list is complete with a helper script?

mattip replied:
I think this might work `runtests.py ... -- --deselect f2py`

DWesl replied:
```bash
for name in numpy/*; 
do 
    if [ -d "${name}" -a "${name}" != numpy/f2py -a "${name:6:1}" != "_" ] ;
    then
        echo ${name##numpy/}; 
    fi;
done
```
or
```bash
find numpy/* -maxdepth 0 -type d -a ! -name f2py -a ! -name _\* | cut -f2 -d/
```
would also work, but I don't think this separation is needed anymore.

DWesl replied:
All submodule tests run at the same time again; PR title changed to match.

mattip reviewed (commented):
LGTM, unbreaks the cygwin build. I am a little concerned about listing the non-f2py directories to test, but maybe this is good enough for now?

DWesl committed:
CI: Unsplit the Cygwin tests.

Ideally this works nicely and I can change the PR name.
If not, I put the split back, then the parallelization if that still doesn't work.

DWesl commented on numpy/f2py/tests/util.py:
```diff
@@ -30,6 +31,9 @@
 _module_num = 5403
 
+if sys.platform == "cygwin":
+    _module_list = []
```
If we want the tests to run with `python runtests.py`, rather than `pip install -e .; /usr/bin/rebase --database $(find numpy -name \*.dll); python runtests.py -n`, I'd want to add the rebase step in `runtests.py` and start this list with all the extension modules in numpy.

mattip replied:
The `runtests.py` workflow is going away with the move to meson, so the `pip install` workflow is fine.

DWesl committed:
CI: Rebase numpy DLLs in runtests.py.

This assumes NumPy is rebased before tests run,
but does not assume the locations are in the database.

seberg reviewed (approved):
This looks OK to me, not super pretty, but also reasonably little additional code in the test.

Thanks for so valiantly tracking this down!

mattip commented:
Thanks @DWesl
